### PR TITLE
add answers to exercise 1-5 chapter 7

### PR DESCRIPTION
**Pull request by ryblovAV**
Hi @axel22, 
Pls see this answers

### CODE REVIEW

File: src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def first(implicit txn: InTxn): P = rFirst.single()
+
+    def first_=(x: P)(implicit txn: InTxn) = rFirst.single.transform(old => x)
+
+    def second(implicit txn: InTxn): Q = rSecond.single()
+
+    def second_=(x: Q)(implicit txn: InTxn) = rSecond.single.transform(old => x)
+
+    def swap()(implicit e: P =:= Q, txn: InTxn): Unit = {
+      val old = Ref[P](first)
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala]: It shouldn't be necessary to create a `Ref` object just to hold the value of the first field.
It is sufficient to read it via `apply` (i.e. just call `first`) and store the value of type `P` directly to the local variable.

    val old = first

File: src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      first = second.asInstanceOf[P]
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala]: Here you can use the `apply` method of `e` to convert directly:

    first = e(second)

File: src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  val p = new TPair[String,String]("first value","second value")
+
+  def swapOne = atomic { implicit txn =>
+    p.swap
+
+    val vF = p.first
+    val vS = p.second
+
+    Txn.afterCommit { _ =>
+      if (vS == vF) log("ERROR !!!!")
```
Comment (axel22) [inline src/main/scala/org/learningconcurrency/exercises/ch7/ex1.scala]: You can use `assert` instead of a `log` here.

**Comment by axel22** [comment]
Thanks, looking good overall!

I only left some minor comments for exercise 1.

**Comment by ryblovAV** [comment]
@axel22 
Thanks for your review, I'll try to fix

**Comment by ryblovAV** [comment]
Hi @axel22,
Pls, see my changes

I use `asInstanceOf[P]` method to convert Q => P.
Pls suggest me, does a better way exist ?

**Comment by axel22** [comment]
Unfortunately, not that I know of. Ideally, `=:=` class should have a `reverse` method that swaps the type parameters. Alternatively, you could require implicit `=:=` in both directions, but I guess this is good as it is!

LGTM

**Comment by axel22** [comment]
Thanks for your contribution!

**Comment by ryblovAV** [comment]
Thanks

**Pull request merged**